lsp/nls/src/field_walker.rs

```diff
@@ -234,36 +293,67 @@ impl FieldDefs {
                 server
                     .cache
                     .get_ref(*file_id)
-                    .map(|term| FieldDefs::resolve(term, &env, server))
+                    .map(|term| FieldDefs::resolve_term(term, &env, server))
                     .unwrap_or_default()
             }
-            Term::Op2(BinaryOp::Merge(_), t1, t2) => {
-                FieldDefs::resolve(t1, env, server).merge_from(FieldDefs::resolve(t2, env, server))
-            }
+            Term::Op2(BinaryOp::Merge(_), t1, t2) => FieldDefs::resolve_term(t1, env, server)
+                .merge_from(FieldDefs::resolve_term(t2, env, server)),
             Term::Let(_, _, body, _) | Term::LetPattern(_, _, _, body) => {
-                FieldDefs::resolve(body, env, server)
+                FieldDefs::resolve_term(body, env, server)
             }
             Term::Op1(UnaryOp::StaticAccess(id), term) => {
-                FieldDefs::resolve_path(term, &[id.ident()], env, server)
+                FieldDefs::resolve_term_path(term, &[id.ident()], env, server)
             }
             Term::Annotated(annot, term) => {
-                // We only check the annotated contracts, not the annotated type.
-                // (Type-based information comes from the inferred types, which
-                // already account for annotated types.)
-                let terms = annot.contracts.iter().filter_map(|ty| {
-                    // TODO: support Dict and Record types also. For now, this is
-                    // enough for completion on basic contract annotations like
-                    // (x | { foo | Number }).fo
-                    if let TypeF::Flat(rt) = &ty.typ.typ {
-                        Some(rt)
+                let defs = annot
+                    .contracts
+                    .iter()
+                    .chain(annot.typ.iter())
+                    .map(|lty| FieldDefs::resolve_type(&lty.typ, env, server));
+                defs.fold(FieldDefs::resolve_term(term, env, server), |acc, def| {
+                    acc.merge_from(def)
+                })
+            }
+            _ => Default::default(),
+        };
+
+        let typ_fields = if let Some(typ) = server.lin_registry.get_type(rt) {
+            log::info!("got inferred type {typ:?}");
+            FieldDefs::resolve_type(typ, env, server)
+        } else {
+            FieldDefs::default()
+        };
+
+        term_fields.merge_from(typ_fields)
```

Comment:
Just an idle observation: `merge_from` could be another good use case for our `Combine` trait.